### PR TITLE
Feature/top heaviness metric

### PR DESCRIPTION
**Pull request by jiacheng-atmos**
**Description**
A new POD - top-heaviness metric is introduced. 

**How Has This Been Tested?**
Please describe the tests that you ran to verify your changes in enough detail that  
someone can reproduce them. Include any relevant details for your test configuration  
such as the Python version, package versions, expected POD wallclock time, and the   
operating system(s) you ran your tests on.

**Checklist:**
- [x] I have reviewed my own code to ensure that if follows the [POD development guidelines](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/dev_guidelines.html)
- [x] The script are written in Python 3.6 or above (preferred; required if funded by a CPO grant), NCL, or R
- [x] All of my scripts are in the diagnostics/ subdirectory, and include a main_driver script template html, and settings.jsonc
- [x] The main_driver script header has all of the information in the POD documentation, excluding the "More about this diagnostic" section
- [x] The POD directory and html template have the same short name as my POD
- [x] The html template has a 1-paragraph synopsis of the POD and links to the main documentation
- [ ] If applicable, I've added a .yml file to src/conda, and my environment builds with conda_env_setup.sh 
- [x] The POD scripts do not access the internet or networked resources
- [x] I have commented the code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation in the POD's doc/ subdirectory
- [x] I have created the directory input_data/obs_data/[pod short name]
- [x] My code is portable; it uses [MDTF environment variables](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_envvars.html), and does not contain hard-coded file or directory paths
- [x] I have added information about the raw data files to the documentation
- [] I have provided the code to generate digested data files from raw data files
- [x] Each digested data file generated by the script contains numerical data (no figures), and is 3 GB or less in size
- [x] The repository contains no extra test scripts or data files
- [x] My branch is up-to-date with the NOAA-GFDL develop branch, and all merge conflicts are resolved

### CODE REVIEW

File: diagnostics/top_heaviness_metric/settings.jsonc [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+  "settings" : {
+    "driver" : "top_heaviness_metric.py",
+    "long_name" : "Top Heaviness Metric Diagnostic",
+    "realm" : "atmos",
+    "description": "The vertical profiles of diabatic heating have important implications for large-scale dynamics, especially for the coupling between the large-scale atmospheric circulation and precipitation processes. We adopt an objective approach to examine the top-heaviness of vertical motion, which is closely related to the heating profiles and a commonly available model output variable. The diagnostic/metric can also be used to evaluate the top-heaviness of diabatic heating.", 
+    "runtime_requirements": {
```
Comment (wrongkindofdoctor) [inline diagnostics/top_heaviness_metric/settings.jsonc:7]: The ',' after "python3": [], is a syntax error. There also seems to be an issue with spacing somewhere that is not immediately clear from visual inspection.  I copied the file contents into the json parser [here](https://jsonformatter.org/), removed the comma, and clicked the format/beautify button to fix the file.

**Comment by wrongkindofdoctor** [comment]
top_heaviness_ratio.py is empty. The missing code should be added back to the file, or the file should be deleted if you don't want to include it.

File: diagnostics/top_heaviness_metric/settings.jsonc (hunk 1 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/settings.jsonc:7]: Many thanks for giving the feedbacks. I have just fixed the issues :).

**Comment by jiacheng-atmos** [comment]
Thank you for your reminder. I have just fixed the issues and fully rewrite
the Top-heavniess POD.

With regards,
Jiacheng

Jess ***@***.***> 于2021年5月26日周三 上午10:19写道：

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+Open source copyright agreement
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The MDTF framework is distributed under the LGPLv3 license (see LICENSE.txt).
+
+Functionality
+-------------
+
+The currently package consists of following functionalities:
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:27]:
```suggestion
The current package consists of following functionalities:
```

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+(1) Calculation of the fractional variance of vertical velocity at each grid point explained by two base functions (i.e., idealized deep convection profile, idealized deep stratiform profile)
+
+(2) Calculation of the top-heaviness ratio (O2/O1)
+
+(3) Other analysis (to be added soon)
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:33]: Can you be a bit more specific?   Are these "other analyses" planned for this current phase of the MDTF?

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+(**) cropping.py can be referenced if code is needed to either shift the grid of your data
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:35]: Is this a horizontal grid shift?

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+Required programming language and libraries
+-------------------------------------------
+
+Python3 packages: "netCDF4", "xarray", "numpy", "scipy", "matplotlib", "basemap"
+
+Required model output variables
+-------------------------------
+
+3-D spatial dimension Omega (units: Pa/s), which can be either the monthly mean in a certain year or the long-term monthly mean (or seasonal) mean.
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:49]: Please provide CMOR variable names of the required fields.

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+
+References
+----------
+
+.. _ref-Muñoz1:
+
+Back, L. E., Hansen, Z., & Handlos, Z. (2017). Estimating vertical motion profile top-heaviness: Reanalysis compared to satellite-based observations and stratiform rain fraction. Journal of the Atmospheric Sciences, 74(3), 855-864. https://doi.org/10.1175/JAS-D-16-0062.1
+
+Jiacheng and Zhuo's paper is under developing...
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:59]: Please provide a proper "in preparation" reference.

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+More about this diagnostic
+--------------------------
+
+Q1 and Q2 (Figure 1a) are two prescribed base functions. Following Back et al. (2017), Q1 as a half sine function, and Q2 as a full sine function, which represent the idealized deep convection profile and the idealized deep stratiform profile, respectively. The vertical velocity can be approximated by Q1 and Q2:
+ω'(x,y,p) = O1(x,y,t) * Q1(p) + O2(x,y,t)*Q2(p)
+
+
+Holding O1 as positively defined, when the ratio of r=O2/O1 increases from -1 to 1, ω' transitions from a bottom-heavy profile to a top-heavy profile (Figure 1b).
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:68]: What do O1 and O2 physically represent?

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+.. figure:: Q1&Q2_R.png
+   :align: center
+   :width: 75 %
+   
+
+   Figure 1. Left: Q1 and Q2; Right: Vertical motion profiles constructed from varying top-heaviness ratio (r; r=-1: dark blue, r=1: dark red).
+   
+
+.. figure:: R2_Between_Recon_Omega&Original.png
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:79]: Could you comment more generally somewhere in this documentation about potential uses of this diagnostic for model development?  For example, the relatively lower r^2 values over land suggest to me that this diagnostic should only be looked at over oceanic regions.  Is that true?

File: diagnostics/top_heaviness_metric/top_heaviness_ratio_calculation.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
+# 28 May top_heaviness_ratio_calculation.py
+#Python packages/ modules imported for the diagnostic
+# the sample monthly data is from ERA5 in July from 2000 to 2019 
+import os
+import xarray as xr
+import numpy as np
+from scipy import integrate
+import matplotlib.pyplot as plt
+from mpl_toolkits.basemap import Basemap
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/top_heaviness_ratio_calculation.py:9]: Basemap is no longer supported.  Please rewrite the plotting using another package, such as Cartopy.

File: diagnostics/top_heaviness_metric/top_heaviness_ratio_robustness_calc.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
+# Last update: 30 May top_heaviness_ratio_robustness_calc.py
+import os
+import xarray as xr
+import numpy as np
+import scipy 
+from scipy import interpolate
+from scipy import integrate
+import matplotlib.pyplot as plt
+from mpl_toolkits.basemap import Basemap
```
Comment (jkrasting) [inline diagnostics/top_heaviness_metric/top_heaviness_ratio_robustness_calc.py:9]: Basemap is no longer supported.  Please rewrite the plotting using another package, such as Cartopy.

**Review by jkrasting — CHANGES_REQUESTED** [review]
Thanks @jiacheng-atmos!   Looks really interesting.

Big issue though --- `basemap` is not longer being developed and supported by the community.  Can you please switch to a more up-to-date mapping package, such as `cartopy`?

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 2 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst]: I'm truly sorry for the late response.  I have just corrected it.

Thank you~

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 3 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst]: After having a conversation with Zhuo Wang, we agree to just have step 1) and 2).

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 4 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst]: Sorry, I accidentally copied this sentence from the example. I have deleted it. There is no problem with our grid data.

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 5 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst]: Thank you for your suggestion. We just made a revision.

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 7 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:62]: You can view Q1 (idealized deep convection profile) and Q2 (idealized deep stratiform profile) as two leading EOF modes if you applied EOF analysis to the velocity field. Then the O1 and O2 are just two corresponding PC time series. The ratio of O2 to O1 can tell us the relative importance of Q1 and Q2.

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 8 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:76]: For 1) We will add more discussion on explaining R2. 
For 2) Yes, what you think is correct. Approximating the vertical motion field with Q1 and Q2 is only valid over the ocean. The situation over the land is more complex.

File: diagnostics/top_heaviness_metric/top_heaviness_ratio_calculation.py (hunk 9 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/top_heaviness_ratio_calculation.py]: We have updated our code with the Cartopy module.

File: diagnostics/top_heaviness_metric/top_heaviness_ratio_robustness_calc.py (hunk 10 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/top_heaviness_ratio_robustness_calc.py]: We have updated the code with the cartopy module.

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 8 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:76]: About "If applicable, I've added a .yml file to src/conda, and my environment builds with conda_env_setup.sh",  may I ask how to check my environment?

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 7 above)
Comment (jkrasting) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:62]: Great thanks @jiacheng-atmos - can you add this to the documentation?

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 8 above)
Comment (jkrasting) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:76]: Should values over land be masked out?

Regarding the conda environment:  @wrongkindofdoctor or @tsjackson-noaa, can you provide some guidance?

**Review by jkrasting — CHANGES_REQUESTED** [review]

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 7 above)
Comment (jiacheng-atmos) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:62]: Sure, definitely

**Review by jkrasting — APPROVED** [review]

File: diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst (hunk 8 above)
Comment (wrongkindofdoctor) [reply diagnostics/top_heaviness_metric/doc/top_heaviness_metric.rst:76]: @jiacheng-atmos Do you mean verify that your environment was installed after running conda_env_setup.sh? If yes, then `conda env list` will show the environments. `conda list` will show the packages+versions in the active environment.

**Comment by jiacheng-atmos** [comment]
Hi, I have a tiny question about the failure alert reported by MDTF_test: 
Here shows the diagnostic result:

> Run cd ../
> Cloning into 'mdtf_test_data'...
> Traceback (most recent call last):
>   File "./mdtf_test_data/mdtf_synthetic.py", line 7, in <module>
>     import pytest
> ModuleNotFoundError: No module named 'pytest'
> Error: Process completed with exit code 1.

I don't fully understand why the program reported that pytest module is missing since my POD does not include it.

Sorry for the inconvenience

**Comment by wrongkindofdoctor** [comment]
@jiacheng-atmos  We've added some updates to the workflow files recently, so syncing your branch with the develop branch should solve the issue (specifically, you need the latest versions of src/conda/_env_synthetic_data.yml, tests/github_actions_test*.jsonc, and .github/workflows/mdtf_test.yml).

**Comment by wrongkindofdoctor** [comment]
@jiacheng-atmos update--it turns out that pytest is missing from the workflow environment yaml. I just submitted PR #249 with the fix, so once it's merged in, go ahead and sync your branch with develop.

**Comment by jiacheng-atmos** [comment]
> @jiacheng-atmos update--it turns out that pytest is missing from the workflow environment yaml. I just submitted PR #249 with the fix, so once it's merged in, go ahead and sync your branch with develop.

**Comment by jiacheng-atmos** [comment]
> @jiacheng-atmos update--it turns out that pytest is missing from the workflow environment yaml. I just submitted PR #249 with the fix, so once it's merged in, go ahead and sync your branch with develop.

Great thanks! I have just updated my develop branch.

**Review by aradhakrishnanGFDL — CHANGES_REQUESTED** [review]
@jiacheng-atmos Thank you for your contributions. 

The testing routines seem to have caught a formatting error in the settings.jsonc. Please double check and provide an updated settings.jsonc, as needed. 

diagnostics/top_heaviness_metric/settings.jsonc 
https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/3c09931014a2097831d2c208a3c58f39c45ea2b9/diagnostics/top_heaviness_metric/settings.jsonc#L11

Please check the data fields and also ensure the closing braces are all provided at the end of the jsonc. I find the online json linter useful for these purposes. 

ERROR: Couldn't parse the settings.jsonc file for POD 'top_heaviness_metric': Syntax error in 'settings.jsonc': Expecting ',' delimiter: line 12 column 16 (char 783).
[32](https://github.com/NOAA-GFDL/MDTF-diagnostics/runs/5456092000?check_suite_focus=true#step:10:32)

**Comment by jiacheng-atmos** [comment]
> @jiacheng-atmos Thank you for your contributions.
> 
> The testing routines seem to have caught a formatting error in the settings.jsonc. Please double check and provide an updated settings.jsonc, as needed.
> 
> diagnostics/top_heaviness_metric/settings.jsonc
> 
> https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/3c09931014a2097831d2c208a3c58f39c45ea2b9/diagnostics/top_heaviness_metric/settings.jsonc#L11
> 
> Please check the data fields and also ensure the closing braces are all provided at the end of the jsonc. I find the online json linter useful for these purposes.
> 
> ERROR: Couldn't parse the settings.jsonc file for POD 'top_heaviness_metric': Syntax error in 'settings.jsonc': Expecting ',' delimiter: line 12 column 16 (char 783). [32](https://github.com/NOAA-GFDL/MDTF-diagnostics/runs/5456092000?check_suite_focus=true#step:10:32)

Thank you for your advice. I have updated the .jsonc file. Issues related to that seem to be resolved by now. Additionally, may I ask how to provide the code to generate digested data files?
"I have provided the code to generate digested data files from raw data files"

**Comment by jiacheng-atmos** [comment]
> Thank you for your advice. I have updated the .jsonc file. Issues related to that seem to be resolved by now.

**Comment by jiacheng-atmos** [comment]
> @jiacheng-atmos Thank you for your contributions.
> 
> The testing routines seem to have caught a formatting error in the settings.jsonc. Please double check and provide an updated settings.jsonc, as needed.
> 
> diagnostics/top_heaviness_metric/settings.jsonc
> 
> https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/3c09931014a2097831d2c208a3c58f39c45ea2b9/diagnostics/top_heaviness_metric/settings.jsonc#L11
> 
> Please check the data fields and also ensure the closing braces are all provided at the end of the jsonc. I find the online json linter useful for these purposes.
> 
> ERROR: Couldn't parse the settings.jsonc file for POD 'top_heaviness_metric': Syntax error in 'settings.jsonc': Expecting ',' delimiter: line 12 column 16 (char 783). [32](https://github.com/NOAA-GFDL/MDTF-diagnostics/runs/5456092000?check_suite_focus=true#step:10:32)

This issue has been resolved

**Comment by wrongkindofdoctor** [comment]
@jiacheng-atmos As with the PR #250, I've switched the target branch to main. I'll ping @jkrasting for a followup review, and we'll see we can get this merged soon. Please provide any observational datasets required for this POD by sharing the link to the server location, or emailing me a tarball with the files.

**Comment by aradhakrishnanGFDL** [comment]
> @aradhakrishnanGFDL 1.If you mean a modified runtime default_tests.jsonc file, it would be useful to have a copy called [pod name]_runtime.jsonc with the sample datasets and settings used for testing in the diagnostics/[pod name] directory 

Yes. Let's discuss this at mini-leads and add it to the checklist. For this POD, @jiacheng-atmos , if you have the runtime default_tests.jsonc you used for your tests, please consider (not a requirement at this time) pushing a copy per @wrongkindofdoctor's suggested path/filename. 

2. The POD README file is probably a better place for required dataset documentation than the code comments 

Great. @jiacheng-atmos can you please update the README based on the above note? 

4. The script to generate the pre-digested obs data can be placed in the diagnostics/[pod name] directory

@jiacheng-atmos Since you asked about the pre-digested obs data generation script locations, please push it to diagnostics/[pod name] directory if it's not there already. 

Many thanks for your contributions.

**Review by wrongkindofdoctor — APPROVED** [review]

**Pull request merged**